### ISSUE TITLE
Microsoft.NET.Test.Sdk v16.7.0 and later are ~4x slower than v16.6.1

### ISSUE THREAD

**Issue by 0xd4d** [microsoft/vstest#2623]
## Description

Microsoft.NET.Test.Sdk v16.7.0 and later are ~4x slower than v16.6.1

I reported https://github.com/microsoft/vstest/issues/2215 a year ago, maybe these issues are related. That issue was fixed and I closed it.

What changed between 16.6.1 and 16.7.0? Serializing more data? Used new serializer version/nuget? Or something else?

## Steps to reproduce

.NET: 5.0.100

ClassLibrary1.csproj:

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <!-- 16.6.1 or 16.9.0-preview-20201020-06 -->
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.0-preview-20201020-06" />
    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.console" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
  </ItemGroup>

</Project>
```

Class1.cs:

```C#
using System.Collections.Generic;
using Xunit;

namespace ClassLibrary1
{
    public class Class1
    {
        [Theory]
        [MemberData(nameof(TestData))]
#pragma warning disable xUnit1026 // Theory methods should use all of their parameters
        public void Test1(int value)
#pragma warning restore xUnit1026 // Theory methods should use all of their parameters
        {
        }
        public static IEnumerable<object[]> TestData {
            get {
                for (int i = 0; i < 100_000; i++)
                    yield return new object[] { i };
            }
        }
    }
}
```

repro:

- Run the commands below
- Change the `Microsoft.NET.Test.Sdk` version: 16.6.1 <--> 16.9.0-preview-20201020-06
- Run the commands below again

```pwsh
dotnet build -v:m -c Release ClassLibrary1.csproj
Measure-Command { dotnet test -c Release -f net5.0 --no-build ClassLibrary1.csproj | Out-Default }
```

Version | Time | Verdict
-------|----------|--------
16.6.1 | 18.9 seconds | ✔️
16.9.0-preview-20201020-06 | 65.7 seconds | ❌

## Environment

Windows 10 20H2

Ubuntu 20.04 (WSL2)

**Comment by nohwnd** [comment]
@Sanan07  @cvpoienaru  you are focusing on perf, could you have a look on this. That's a huge downgrade in performance.

**Comment by Sanan07** [comment]
@0xd4d I tested following your instructions and got these results:

When I test using .NET SDK 3.1.403 I got the same numbers as yours and sometimes a bit higher (around 4 minutes), but when I use the latest .NET SDK 5.0.100-rc.2.20479.15 I got :

Version | Time
--------|------
16.6.1 | 1 m 46 s 
16.7.0 | 2 m
16.9.0-preview-20201020-06 |  1 m 47 s 

These are screenshots : 
![16 6 1](https://user-images.githubusercontent.com/8465726/98546196-db066400-2296-11eb-82bb-bd5297d410e1.PNG)
![16 7 0](https://user-images.githubusercontent.com/8465726/98546199-dc379100-2296-11eb-8d85-b649ad49c43c.PNG)
![16 9 0](https://user-images.githubusercontent.com/8465726/98546201-dc379100-2296-11eb-8c29-b584f390d0da.PNG)

Could you please also install .NET SDK 5.0 and test it and see if you also get less time?

**Comment by 0xd4d** [comment]
.NET SDK: 5.0.100-rc.2.20479.15

version | VS test says | actual time measured with `Measure-Command`
--------|--------------|-------------------------------------
16.6.1 | Duration: 1 m 1 s | 58.8492535 seconds
16.9.0-preview-20201020-06 | Duration: 57 s | 210.8753572 seconds

It claims to have executed in 57 s but the actual time is 211s.

With 16.6.1 it claims to have executed in 61s but the actual time was 59s...

```pwsh
Measure-Command { orig-command-here | Out-Default }
```

Try this with .NET 5.0.100-rc.2.20479.15

```pwsh
git clean -xdf
dotnet build -v:m -c Release src/csharp/Iced.sln
Measure-Command { dotnet test -c Release -f netcoreapp3.1 --no-build src/csharp/Intel/Iced.UnitTests/Iced.UnitTests.csproj -- RunConfiguration.NoAutoReporters=true RunConfiguration.TargetPlatform=X64 | Out-Default }
```

**Comment by Sanan07** [comment]
Yes, I agree with you about that numbers. I tested using your `Measure-Command` and also in parallel started Stopwatch in Windows. So the numbers of `Measure-Command` and Stopwatch were the same, but `dotnet test` number was 2 times less.

![meas](https://user-images.githubusercontent.com/8465726/98679069-bd510180-235f-11eb-900d-efb9e144bd44.PNG)

Something wrong with our duration calculation. Potentially can be race condition issue. We use ConcurentDictionary for inserting duration numbers which does not guarantee full thread safety, so need to look at it as well. Or maybe somehow related to this #2585 

We will come back to you, if we get some new investigation results.

**Comment by nohwnd** [comment]
@Sanan07 what you measure by the wall clock (Measure-Command), is all startup cost, discovery and all other overhead. The time reported by the summary is the time spent running tests.

**Comment by 0xd4d** [comment]
As I suspected, it appears to be a serialization issue. I've updated the first post with a minimal repro.

I debugged the code and randomly saved 2 serialized messages received by `TestRequestSender.OnExecutionMessageReceived()`, here are the results (length in chars):

 Length|Name|Version
 ------|----|-------
  30453|new-msg|16.9.0-preview-20201020-06
  30454|new-msg2|16.9.0-preview-20201020-06
   7224|old-msg|16.6.1
   7228|old-msg2|16.6.1

16.6.1 uses version 2 and 16.9.0 uses version 3. Version 3 has a lot of useless stuff that makes the json 4x as big as the old version.

```json
"TestCase": {
	"Id": "410fae12-2d5b-4097-bb5b-2847ae07c593",
	"FullyQualifiedName": "ClassLibrary1.Class1.Test1",
	"DisplayName": "ClassLibrary1.Class1.Test1(value: 87453)",
	"ExecutorUri": "executor://xunit/VsTestRunner2/netcoreapp",
	"Source": "C:\\testing\\ClassLibrary1\\ClassLibrary1\\bin\\Debug\\net5.0\\ClassLibrary1.dll",
	"CodeFilePath": null,
	"LineNumber": 0,
	"Properties": []
},
```

```json
"TestCase": {
	"Properties": [
		{
			"Key": {
				"Id": "TestCase.FullyQualifiedName",
				"Label": "FullyQualifiedName",
				"Category": "",
				"Description": "",
				"Attributes": 1,
				"ValueType": "System.String"
			},
			"Value": "ClassLibrary1.Class1.Test1"
		},
		{
			"Key": {
				"Id": "TestCase.ExecutorUri",
				"Label": "Executor Uri",
				"Category": "",
				"Description": "",
				"Attributes": 1,
				"ValueType": "System.Uri"
			},
			"Value": "executor://xunit/VsTestRunner2/netcoreapp"
		},
		{
			"Key": {
				"Id": "TestCase.Source",
				"Label": "Source",
				"Category": "",
				"Description": "",
				"Attributes": 0,
				"ValueType": "System.String"
			},
			"Value": "C:\\testing\\ClassLibrary1\\ClassLibrary1\\bin\\Debug\\net5.0\\ClassLibrary1.dll"
		},
		{
			"Key": {
				"Id": "TestCase.CodeFilePath",
				"Label": "File Path",
				"Category": "",
				"Description": "",
				"Attributes": 0,
				"ValueType": "System.String"
			},
			"Value": null
		},
		{
			"Key": {
				"Id": "TestCase.DisplayName",
				"Label": "Name",
				"Category": "",
				"Description": "",
				"Attributes": 0,
				"ValueType": "System.String"
			},
			"Value": "ClassLibrary1.Class1.Test1(value: 16692)"
		},
		{
			"Key": {
				"Id": "TestCase.Id",
				"Label": "Id",
				"Category": "",
				"Description": "",
				"Attributes": 1,
				"ValueType": "System.Guid"
			},
			"Value": "4ed1440e-a5a6-409c-af9a-27410afd3bb4"
		},
		{
			"Key": {
				"Id": "TestCase.LineNumber",
				"Label": "Line Number",
				"Category": "",
				"Description": "",
				"Attributes": 1,
				"ValueType": "System.Int32"
			},
			"Value": 0
		}
	]
},
```

**Comment by Sanan07** [comment]
@0xd4d Thank you for your investigation. We also found out this. So I think we know how to fix this and after we do it, we will put an update here.

**Comment by nohwnd** [comment]
@0xd4d the issue was that there is a custom serializer used, and when protocol version was updated to 3, it started using serializer for protocol version 1, which produces large messages. I forced Test.SDK to dowgrade to protocol 2 when asked for 3, and pushed the current version to 4, which will use the correct serializer. This should fix the speed issue for older console version which will only ever ask for version 3, but won't be updated to ask for 4.
